All filter acting wrong. #2565

dbeltrankyl opened the pull request:
closes https://github.com/BSC-ES/autosubmit/issues/2565.

Part of https://github.com/BSC-ES/autosubmit/issues/2666

<!--

Thanks for your contribution! Please:
* List any related issues with a "closes" or "addresses" tag.
* Add a helpful title & description.
* Complete the checklist.

-->
closes #2565.
**Check List**

- [x] I have read `CONTRIBUTING.md`.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to `pyproject.toml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included in `CHANGELOG.md` if this is a change that can affect users.
- [x] Documentation updated.
- [x] If this is a bug fix, PR should include a link to the issue (e.g. `Closes #1234`).

dbeltrankyl commented:
Ready to review, just to mention that I did the less invasive changes.

kinow commented on autosubmit/job/job_list.py:
```diff
@@ -353,23 +353,52 @@ def _deep_map_dependencies(self, section, jobs_data, option, dependency_list=set
+            self,
+            date_list: List[Any],
+            member_list: List[Any],
+            chunk_list: List[int],
```
s/List/list 😬

kinow reviewed (commented):
@dbeltrankyl I was looking inside `_deep_map_dependencies`, and noticed that they also strip keys there, so I thought we could re-use the new function you wrote for that, reducing duplicated code.

But, there's something that I don't understand in that function.

It takes a `strip_keys: bool`, defaulting to `True`. Which is fine. I take it it defines whether you want to strip keys or not.

It ignores and returns immediately if section already in list, which makes sense (recursivity), or adds the section to the list.

The next step is then to check, if `strip_keys` is **`False`**, then it checks if the section contains `+` or `-`, and then does the `section.split` (i.e. what your `_strip_key` function is doing, I think).

Then, for each dependency of the section it checks if `strip_keys` is **`True`**, hen it calls the `dependency.split`, which I thought would be doing the same thing as your function.

Do you know why first `strip_keys` being `False` triggers the code to split by `+`, `-`, and then in the same code if `strip_keys` is `True` is what triggers that again? 😵

dbeltrankyl commented:
@kinow

> Do you know why first strip_keys being False triggers the code to split by +, -, and then in the same code if strip_keys is True is what triggers that again? 😵

Yeah, it's confusing, and there should be a better way to do it, for sure. 

Context:

The outcome of the function should be this: 

What this is supposed to do is navigate through 

JOBS.JOB.DEPENDENCY

These keys can have distances ( sim-1, sim+1) 

There are two maps, one with distances and another one without them

The one without distances is straightforward; strip all the keys always

ex:

current_section:
map["section"] = ("other_section")
Then we want to navigate to other_section
map["other_section"] = ("other_other_section")

and so on 

The one with distances, 

current_section:
map_d["section"] = ("other_section-1")
Then we want to navigate to other_section, but we need to strip the other_section for the key name
map_d["other_section"] = ("other_other_section",)

---------------------------------

Debug:

The first call (i=0), always has the section stripped as it comes from the jobs_data

So, this can be ignored for i=0

```python
        if not strip_keys:
            if "+" in section:
                section = section.split("+")[0]
            elif "-" in section:
                section = section.split("-")[0]
```

This one is for the dependencies, here we can have +/-

```python
        for dependency in dependencies_keys:
            if strip_keys:
                if "+" in dependency:
                    dependency = dependency.split("+")[0]
                elif "-" in dependency:
                    dependency = dependency.split("-")[0]
```

map: i=1: new_section= dependency_stripped 
map_distances: i=1: the new_section=dependency can have +/-

So, this in map_distances case i=1

```python
        if not strip_keys:
            if "+" in section:
                section = section.split("+")[0]
            elif "-" in section:
                section = section.split("-")[0]
```

will strip the key from the section, so we can find it in the jobs_data dict ( jobs.job)

kinow reviewed (approved):
Added a comment about a typo, and types that can use `list` instead of `List`. I don't know if your new small function to strip keys can be used in the other function, Dani, but that can be done later as well. Feel free to merge as is or fix anything of these small issues if you have time. Thanks!

kinow commented:
(Moved the closes #issuenumber to main comment as I **think** it only works there?)
